test/cpp/api/inference_mode.cpp

```diff
@@ -0,0 +1,276 @@
+  ASSERT_TRUE(tmp.requires_grad());
+}
+
+TEST(InferenceModeTest, TestRequiresGradTrueTensorViewOp) {
+  torch::Tensor s = torch::ones({1, 2, 3}).set_requires_grad(true);
+  torch::Tensor a = s + 2;
+  torch::Tensor view_out, tmp;
+
+  {
+    c10::InferenceMode guard;
+    // View ops on normal tensor produce normal tensors as output.
+    // - For view ops it has both dispatch keys since due to the way we create
+    //   view Tensors in alias_with_sizes_and_strides:
+    //   ```
+    //     auto impl = c10::make_intrusive<TensorImpl>(
+    //     Storage(self.storage()), self.key_set(), self.dtype());
+    //   ```
+    //   In addition, these view output tensors although are normal in the sense
+    //   that it has both Autograd and InplaceOrView keys, they're still special
+    //   since they'll have CreationMeta::NO_VARIABLE_TYPE_VIEW since they don't
+    //   have proper setup for backward node (skipping VariableType kernel).
+    //   In other words, they'll be specially handled if participating autograd
+    //   outside InferenceMode.
+
+    view_out = view_op(a);  // go through kernels: InplaceOrView, CPU
+    assert_tensor_dispatch_keys(view_out, /*is_inference_tensor=*/false);
+    assert_tensor_creation_meta(view_out, torch::autograd::CreationMeta::NO_VARIABLE_TYPE_VIEW);
+    ASSERT_TRUE(view_out.requires_grad());
+    ASSERT_FALSE(view_out.is_leaf());
+
+    // view -> view
+    tmp = view_op(view_out);  // go through kernels: InplaceOrView, CPU
+    assert_tensor_dispatch_keys(tmp, /*is_inference_tensor=*/false);
+    assert_tensor_creation_meta(tmp, torch::autograd::CreationMeta::NO_VARIABLE_TYPE_VIEW);
+    ASSERT_TRUE(tmp.requires_grad());
+    ASSERT_FALSE(tmp.is_leaf());
+
+    // view -> view -> inplace
+    inplace_op(tmp);  // kernels: InplaceOrView, CPU
+    assert_tensor_creation_meta(tmp, torch::autograd::CreationMeta::NO_VARIABLE_TYPE_VIEW);
+    assert_tensor_dispatch_keys(tmp, /*is_inference_tensor=*/false);
+    ASSERT_TRUE(tmp.requires_grad());
+  }
+
+  ASSERT_THROWS_WITH(functional_op(view_out), // go through kernels: VariableType, InplaceOrView(fallthrough), CPU
+    "A view created in InferenceMode without proper grad_fn setup")
+
+  ASSERT_THROWS_WITH(inplace_op(view_out),  // go through kernels: VariableType, InplaceOrView, CPU
+    "A view created in InferenceMode without proper grad_fn setup")
+
+  ASSERT_THROWS_WITH(view_op(view_out), // go through kernels: VariableType, InferenceMode, CPU
+    "A view created in InferenceMode without proper grad_fn setup")
+
+  // Suggested workaround
+  torch::Tensor detached = view_out.detach();
+  assert_tensor_dispatch_keys(detached, /*is_inference_tensor=*/false);
+  functional_op(detached);
+  inplace_op(detached);
+  view_op(detached);
+}
+
+TEST(InferenceModeTest, TestMixInferenceAndNormalTensor) {
+  // Note you can only mix inference tensor and normal tensor in functional ops.
+  // FIXME: Inplace & view ops only take one tensor input.
```

Comment:
Inplace ops can take more than one tensor, even `tensor.add_(tensor)` is a simple example